Limit dbuf cache sizes based only on ARC target size by default

freqlabs opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Fixes #10563

Dbuf caches are incorrectly limited to too small max size.

### Description
<!--- Describe your changes in detail -->
Set the initial max sizes to ULONG_MAX to allow the caches to grow
with the ARC.

Recalculate the metadata cache size on demand so it can adapt, too.

Update descriptions in zfs-module-parameters(5).

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->
Built on FreeBSD, passes ZTS

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

freqlabs commented:
I think including arc_impl.h might be a sign I'm breaking some organizational boundary. Maybe a function should be added in arc.h to return arc_c_max, similar to arc_target_bytes().

ahrens commented on module/zfs/dbuf.c:
```diff
@@ -613,8 +614,7 @@ dbuf_cache_multilist_index_func(multilist_t *ml, void *obj)
 {
-	return MIN(dbuf_cache_max_bytes,
-	    arc_target_bytes() >> dbuf_cache_shift);
+	return MIN(dbuf_cache_max_bytes, arc_c_max >> dbuf_cache_shift);
```
This seems like it would still defeat any manual increase to the `dbuf_cache_max_bytes` module parameter.  Is that the intention?  If so then it seems like we should initialize dbuf_cache_max_bytes to ULONG_MAX, and then we don't need to do anything with it in `dbuf_init()` (it can be tuned down from there and it will take effect here in dbuf_cache_target_bytes()).

freqlabs replied:
How about `MIN(dbuf_cache_max_bytes, arc_c_max)`? I think the intent here is to allow this to be tunable, but not allow for an unreasonable value.

freqlabs replied:
But on the other hand, initializing `dbuf_cache_max_bytes` to ULONG_MAX makes the dbuf cache target follow changes to `arc_c_max` (via `zfs_arc_max`).

amotin replied:
There is not much difference between setting dbuf_cache_max_bytes to ULONG_MAX and the change below.  Setting it to ULONG_MAX just leaves the question about dbuf_metadata_cache_max_bytes. I remember some requests to remove the dbuf metadata limit, but I don't remember motivation why it was added to vote for removal.

This particular chunk either way I'd leave as it is, as I have told.

amotin commented on module/zfs/dbuf.c:
```diff
@@ -613,8 +614,7 @@ dbuf_cache_multilist_index_func(multilist_t *ml, void *obj)
 {
-	return MIN(dbuf_cache_max_bytes,
-	    arc_target_bytes() >> dbuf_cache_shift);
+	return MIN(dbuf_cache_max_bytes, arc_c_max >> dbuf_cache_shift);
```
Why have you changed this? To me previous dbuf sizing based on ARC sizing looked reasonable.  With this you practically remove back-pressure on dbuf cache when OS asks for RAM back.

freqlabs commented:
I'll need to update zfs-module-parameters(5) with this change, too.

freqlabs commented:
Should I just go ahead and make the metadata limit follow the target arc size as well? ie
```C
static inline unsigned long
dbuf_metadata_cache_target_bytes(void)
{
        return (MIN(dbuf_metadata_cache_max_bytes,
            arc_target_bytes() >> dbuf_metadata_cache_shift));
}
```

amotin commented:
> Should I just go ahead and make the metadata limit follow the target arc size as well?

It may have sense.  IIRC there is no eviction from the dbuf metadata cache at all, so at least making it grow slower together with ARC may be better then nothing from the memory pressure side.  But it was written by @ahrens , so I'll leave it to him to comment.

ahrens commented:
I think it would be OK for the dbuf metadata limit to also scale with arc_c.

ahrens commented on module/zfs/dbuf.c:
```diff
@@ -807,20 +808,14 @@ dbuf_init(void)
-	 * of the target size of the ARC. If the values has been specified as
-	 * a module option and they're not greater than the target size of the
-	 * ARC, then we honor that value.
+	 * dbuf metadata cache to 1/16th (default) of the max size of the ARC.
```
[nit]  It's actually 1/64th (`dbuf_metadata_cache_shift = 6`).  I have a fix for this in another branch that I'll get opened soon, but if you have a mind to fix it now that would be fine too.

freqlabs replied:
Fixed this in the manual. The comment here was removed but I added comments to describe the dbuf cache targets elsewhere, without this error.

amotin commented on man/man5/zfs-module-parameters.5:
```diff
@@ -47,13 +46,12 @@ Default value: \fB0\fR.
-cache and its associated settings can be observed via the
-\fB/proc/spl/kstat/zfs/dbufstats\fR kstat.
+Maximum size in bytes of the metadata dbuf cache.  The target size is
+determined by the MIN versus \fB1/2^dbuf_cache_shift\fR (1/16) of the target
```
dbuf_metadata_cache_shift

freqlabs commented:
* Fixed errors in the tunable descriptions in zfs-module-parameters(5) while touching those lines
* Added comments to dbuf_cache_target_bytes and dbuf_metadata_cache_target_bytes to convey their intent

freqlabs commented:
I don't need to include arc_impl.h anymore.

freqlabs commented:
* Rebased
* Dropped unnecessary header changes
